TranscriptTestRunner/TestRunner.cs

```diff
@@ -48,38 +93,58 @@ private void ConvertTranscript(string transcriptPath)
 
-            var testScript = JsonConvert.DeserializeObject<TestScript[]>(reader.ReadToEnd());
+            var testScript = JsonConvert.DeserializeObject<TestScript[]>(await reader.ReadToEndAsync().ConfigureAwait(false));
 
-            foreach (var script in testScript)
+            foreach (var scriptActivity in testScript)
             {
-               if (script.Role == "bot")
-               {
-                   var activity = new Activity
-                   {
-                       Type = script.Type,
-                       Text = script.Text
-                   };
-
-                   if (!await TestClientBase.ValidateActivityAsync(activity).ConfigureAwait(false))
-                   {
-                       throw new Exception($"The bot didn't reply as expected. It should have said: {activity.Text}");
-                   }
-               }
-               else
-               {
-                   var activity = new Activity
-                   {
-                       Type = script.Type,
-                       Text = script.Text
-                   };
-
-                   await TestClientBase.SendActivityAsync(activity).ConfigureAwait(false);
-               }
+                switch (scriptActivity.Role)
+                {
+                    case RoleTypes.User:
+                        // Send the activity.
+                        var sendActivity = new Activity
+                        {
+                            Type = scriptActivity.Type,
+                            Text = scriptActivity.Text
+                        };
+
+                        await SendActivityAsync(sendActivity, cancellationToken).ConfigureAwait(false);
+                        break;
+
+                    case RoleTypes.Bot:
+                        // Assert the activity returned
+                        if (!IgnoreScriptActivity(scriptActivity))
+                        {
+                            await AssertReplyAsync(
+                                nextReply =>
+                                {
+                                    if (scriptActivity.Type != nextReply.Type)
+                                    {
+                                        throw new Exception($"Invalid activity type. Expected: {scriptActivity.Type} Actual: {nextReply.Type}");
+                                    }
+
+                                    if (scriptActivity.Text != nextReply.Text)
+                                    {
+                                        throw new Exception($"Invalid activity text. Expected: {scriptActivity.Text} Actual: {nextReply.Text}");
+                                    }
+                                },
+                                cancellationToken).ConfigureAwait(false);
+                        }
+
+                        break;
+
+                    default:
+                        throw new InvalidOperationException($"Invalid script activity type {scriptActivity.Role}.");
+                }

```

Comment:
Some activities don't have a Role defined, for example, the _UpdateConversation_. This will fail for those cases. Should we ignore these types of activities?

Reply:
Not sure, are you taking about conversationUpdate?

Reply:
Yes, sorry, _conversationUpdate_. The transcripts we are using contain _conversationUpdate_ activities at the beginning of the conversation.

Reply:
Interesting, we may have a bug somewhere. the original transcripts have role:
![image](https://user-images.githubusercontent.com/12264946/96602216-c864c480-12c0-11eb-94bf-067e2a9ab75c.png)

but the generated script doesn't.

I tested it and it works but also noticed something strange with the ConversationUpdate from the DL client that I need to research a bit more. 

Let's chat on the phone about this one in a bit